File: presentation/src/main/java/by/grodno/vasili/githubsimpleapp/feature/users/UserItemDatasourceFactory.java
package by.grodno.vasili.githubsimpleapp.feature.users;

import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;
import androidx.paging.PageKeyedDataSource;
import by.grodno.vasili.domain.interactor.GetUsersUseCase;

/**
 * Factory for create {@link UserItemDatasource} and store it in LiveData
 */
public class UserItemDatasourceFactory extends DataSource.Factory<String, UserItem> {
    private final GetUsersUseCase getUsersUseCase;
    private final MutableLiveData<PageKeyedDataSource<String, UserItem>> dataSourceLiveData;

    UserItemDatasourceFactory(GetUsersUseCase getUsersUseCase) {
        this.getUsersUseCase = getUsersUseCase;
        dataSourceLiveData = new MutableLiveData<>();
    }

    @Override
    public DataSource<String, UserItem> create() {
        UserItemDatasource datasource = new UserItemDatasource(getUsersUseCase);
        dataSourceLiveData.postValue(datasource);
        return datasource;
    }

    MutableLiveData<PageKeyedDataSource<String, UserItem>> getDataSourceLiveData() {
        return dataSourceLiveData;
    }
}

File: presentation/src/main/java/by/grodno/vasili/githubsimpleapp/feature/users/UserItemDatasourceFactory.kt
package by.grodno.vasili.githubsimpleapp.feature.users

import androidx.lifecycle.MutableLiveData
import androidx.paging.DataSource
import androidx.paging.PageKeyedDataSource
import by.grodno.vasili.domain.interactor.GetUsersUseCase
import kotlinx.coroutines.CoroutineScope

/**
 * Factory for create [UserItemDatasource] and store it in LiveData
 */
class UserItemDatasourceFactory(
        private val getUsersUseCase: GetUsersUseCase,
        private val scope: CoroutineScope
) : DataSource.Factory<String, UserItem>() {
    val dataSourceLiveData: MutableLiveData<PageKeyedDataSource<String, UserItem>> = MutableLiveData()

    override fun create(): DataSource<String, UserItem> {
        val datasource = UserItemDatasource(getUsersUseCase, scope)
        dataSourceLiveData.postValue(datasource)
        return datasource
    }
}
